### ISSUE TITLE
[training/ipc] iPC accumulates weight gradients across all T inference iterations with no 1/T normalization, coupling gradient scale to T and breaking LR transfer to backprop/DFA arms

### ISSUE THREAD

**Issue by harshithkantamneni** [harshithkantamneni/agi-lab-framework#10]
## What's wrong

The "Inference Predictive Coding" (iPC) training routine runs an inner loop over `T` inference iterations per sample (default `T = 5`). Inside that loop it both computes prediction errors *and* accumulates weight gradients. Crucially, the gradient buffers are zeroed only once, before the loop starts (and only on the first micro-batch), and every iteration *adds* to the same buffers (`+=`). So after the loop finishes, each weight's gradient is the sum of `T` per-iteration contributions.

The only place anything gets divided down afterward is the gradient-accumulation averaging in Phase 3, which divides by `accum_steps` (the micro-batch count) — **not** by `T`. There is no `1/T` factor anywhere. The per-element `1/(seq_len*D)` factor used inside Steps C/D is applied *per iteration*, so it scales each contribution but does not remove the fact that `T` of them are summed.

Net effect: with `T = 5`, every iPC weight gradient is carried at roughly a 5x scale compared to a single forward/backward pass. The backprop arm (`backprop_train.c`) and the DFA / local-feedback arm (`local_feedback_train.c`) have **no** such loop — they do a single pass, so their gradients are at ~1x scale. All three arms then hand their gradients to the *same* optimizer entry points (`grad_apply_adam` / `grad_apply_sgd` in `grad.c`), which use the *same* learning-rate schedule (`compute_lr` in `grad.c`). The result is that the effective learning rate the iPC weights actually see is silently multiplied by ~T relative to the other two arms, and it shifts whenever `T` is tuned.

## Why it matters

- **Learning rate is not transferable across arms.** The framework appears designed to compare iPC vs backprop vs DFA under shared infrastructure (shared `grad.c` / `compute_lr`). But "the same LR" does not mean "the same effective step" — the iPC arm's gradients are ~T× larger, so a fair comparison at a given configured LR is actually comparing iPC-at-(LR·T) against backprop-at-LR. Anyone interpreting the comparison at face value can draw wrong conclusions about which training method is better.
- **Hidden coupling to an inference hyperparameter.** `T` is conceptually an *inference / convergence* knob (how many times the value nodes settle). Changing it from 5 to, say, 10 silently doubles the effective gradient magnitude, which can destabilize training or change results in a way that looks like "more inference iterations help/hurt" when it is really just a learning-rate change.
- **Notice point:** whoever sweeps `T`, or whoever tunes one LR and assumes it carries to all three training modes, hits this. It typically shows up as iPC needing a much smaller configured LR than backprop for stable training, with no documented reason.

It is absorbable (you *can* compensate via the LR), which is why this is a design smell rather than an outright bug — but it is an unmodeled, undocumented scale factor.

## Where

`src/training/ipc_train.c`

The T loop opens at line 693 and closes at line 1403:
```c
693:    for (int32_t t = 0; t < T; t++) {
```
Step B accumulates the embedding/output-head gradient every iteration (line 734):
```c
734:        accum_matmul_tn(grads->embed_grad.dweight, grad_logits, normed,
735:                        seq_len, cfg->vocab_size, D);
```
Step D accumulates every per-layer weight gradient every iteration (e.g. lines 816-821), where `accum_matmul_tn` is purely additive (line 234):
```c
234:                dw[i * N + j] += ai * bs[j];
```
Gradients are zeroed only before the loop, first micro-batch only (line 269):
```c
269:        grad_zero(grads);
```
The default is `T = 5` (`src/training/train_config.h:118`, field documented at line 17).

Phase 3 averaging divides by `accum_steps`, never by `T` (lines 1421-1465):
```c
1424:            float inv_accum = 1.0f / (float)accum_steps;
```
Shared optimizer / LR with the other arms: `grad_apply_adam`/`grad_apply_sgd` and `compute_lr` live in `src/training/grad.c` (`compute_lr` at `grad.c:11`, used at `grad.c:285` and `grad.c:536`) and are called by all three arms (`ipc_train.c:1475/1477`, `backprop_train.c:1502/1510`, `local_feedback_train.c:1200/1207`). Neither backprop nor local_feedback has a `t < T` loop (confirmed by grep — no matches).

## How to fix

Pick one of the following; the first is the smallest, safest change.

1. **Normalize iPC weight gradients by `T` so a configured LR means the same thing across all three arms.** Add a single scale pass after the T loop (and before the existing Phase 3 averaging), guarded so it only runs once per micro-batch:

```c
/* After the T loop closes (after line 1403), before Phase 3 averaging.
 * iPC summed weight gradients over T inference iterations; divide by T so
 * the effective LR matches the single-pass backprop/DFA arms. */
if (T > 1) {
    float inv_T = 1.0f / (float)T;
    /* reuse the same SCALE_TENSOR pattern already used for inv_accum below,
     * applying inv_T to embed_grad, final_norm_grad, and every block grad. */
}
```
The cleanest implementation folds `inv_T` into the existing Phase 3 scaling block (lines 1421-1465): change `inv_accum = 1.0f / accum_steps` to a combined factor `inv_scale = 1.0f / ((float)accum_steps * (float)T)` and run that block unconditionally (not only when `accum_steps > 1`), since now there is always a normalization to apply. Be careful: the embedding/output-head grad (`embed_grad.dweight`) is accumulated in Step B every iteration too, so it must be included in the `1/T` scaling exactly like the other tensors.

2. **Alternatively, only accumulate weight gradients on the final iteration** (`if (t == T - 1)`) so the inner loop is purely value-node inference and gradients reflect a single converged pass. This changes the learning dynamics more (it is arguably the more standard iPC formulation) and should be benchmarked against the current behavior before adopting.

Either way, **document the choice** in the comment block at lines 645-653 (currently it lists Steps A-D inside the loop without noting the T-fold gradient accumulation).

Suggested test to add: a small unit/integration check that runs one iPC step with `T = 1` and one with `T = 2` on identical inputs/weights and asserts the resulting weight-gradient L2 norm is (approximately) equal after the fix — today it is ~2× larger for `T = 2`. This pins the invariant that gradient scale is independent of `T`.

## Severity

Medium — it does not crash or corrupt anything and is absorbable by tuning the learning rate, but it silently couples gradient magnitude to an inference hyperparameter and makes the learning rate non-transferable between the iPC arm and the backprop/DFA arms that share the same optimizer and LR schedule, which can quietly bias any cross-method comparison.
